### PR TITLE
Fixed hanging connections. Added viability checker to NWSocket.

### PR DESCRIPTION
**Pull request by mkrn**
- Fixed a situation when socket is hanging while trying to close a non-viable connection
- Fixed a situation when the connection is closed but RTMPConnection is still trying to send data.
 

## Description & motivation
Fixes https://github.com/shogo4405/HaishinKit.swift/issues/1210 

## Type of change
Please delete options that are not relevant.
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: Sources/RTMP/RTMPConnection.swift [hunk 1]
```diff
@@ -456,6 +458,19 @@ open class RTMPConnection: EventDispatcher {
...
+            zeroBytesOutCounter += 1
+            // If currentBytesOutPerSecond has been 0 for the last 5 seconds
+            if zeroBytesOutCounter >= 5 {
+                print("STALE connection, force close")
+                self.close(isDisconnected: true) // call close function
+                zeroBytesOutCounter = 0 // reset counter
+            }
+        } else {
+            zeroBytesOutCounter = 0 // reset counter
+        }
```
Comment (shogo4405) [inline Sources/RTMP/RTMPConnection.swift:472]: I believe it is abnormal for there to be no output for 5 seconds during publishing. On the other hand,  Playback feature it is a possible scenario. If I accept it as it is, it will be considered a bug.

Comment (shogo4405) [reply Sources/RTMP/RTMPConnection.swift:472]: It seems like a good idea to implement the PING/PONG mechanism for RTMP.

File: Sources/RTMP/RTMPNWSocket.swift [hunk 2]
```diff
@@ -103,16 +110,7 @@ final class RTMPNWSocket: RTMPSocketCompatible {
...
-        if connection.state == .ready {
-            outputQueue.async {
-                let completion: NWConnection.SendCompletion = .contentProcessed { (_: Error?) in
-                    self.connection = nil
-                }
-                connection.send(content: nil, contentContext: .finalMessage, isComplete: true, completion: completion)
-            }
-        } else {
-            self.connection = nil
-        }
```
Comment (shogo4405) [inline Sources/RTMP/RTMPNWSocket.swift:115]: This process was created specifically to address this issue, so simply removing it would cause the problem to reoccur.

**Review by shogo4405 — CHANGES_REQUESTED** [review]
I think it would be a good idea to introduce connection?.viabilityUpdateHandler, so I will try.

**Comment by shogo4405** [comment]
Thank you for providing the PR. I will refer to it and try to make the necessary fixes on my end as well.

File: Sources/RTMP/RTMPConnection.swift (hunk 1 above)
Comment (mkrn) [reply Sources/RTMP/RTMPConnection.swift:472]: Yes  maybe ping pong is a better idea, I couldn't tell if it's standard for RTMP.
For me the stale connections happen with this output:

2023-06-22 16:10:31.674537+0100 appnew[1163:121214] [tcp] tcp_input [C15:1] flags=[R] seq=1983586932, ack=0, win=0 state=CLOSE_WAIT rcv_nxt=1983586933, snd_una=1331617881


CLOSE_WAIT means that the local end of the connection has received a FIN from the other end, but the OS is waiting for the program at the local end to actually close its connection. The problem is your program running on the local machine is not closing the socket.

After CLOSE_WAIT notice from NWConnection provided by OS, RTMP Connection is still active and sends publishSufficientBWOccured,  but currentBytesOutPerSecond are 0  and it can last for long time  1+ minutes.

I don't know where CLOSE_WAIT is coming from.. maybe server issue. Tried with different servers,  RTMP, RTMPS, with balancer, and without, but still getting CLOSE_WAIT when reconnecting quickly.

File: Sources/RTMP/RTMPNWSocket.swift (hunk 2 above)
Comment (mkrn) [reply Sources/RTMP/RTMPNWSocket.swift:115]: I see... What was happening is: connection was stale so end message could not be sent..  so it was stuck forever,  completion handler was never reached.

Maybe with viabilityUpdateHandler check for viability this could work

File: Examples/iOS/Preference.swift [hunk 3]
```diff
@@ -1,6 +1,6 @@
 struct Preference {
     static var defaultInstance = Preference()
 
-    var uri: String? = "rtmp://192.168.1.6/live"
-    var streamName: String? = "live"
+    var uri: String? = "rtmps://tempbalancer.eventlive.pro:443/live"
```
Comment (leo150) [inline Examples/iOS/Preference.swift:4]: There are some changes in this commit that are not related to the fix itself. Could you please remove it? This url, new DEVELOPMENT_TEAM, changing maxRetryCount, commenting code, new padding in stateDidChange. Thank you

**Comment by shogo4405** [comment]
I have incorporated the elements of this PR into the main branch. The only remaining task is to properly implement RTMP PING/PONG. You can go ahead and close the PR. Thank you very much.
